README.md
# Cardinal

[![npm version](https://badge.fury.io/js/cardinalcss.svg)](https://badge.fury.io/js/cardinalcss)
[![Bower version](https://badge.fury.io/bo/cardinal.svg)](https://badge.fury.io/bo/cardinal)

Build more with less.

- [Go to Website](http://cardinalcss.com)
- [Review the CHANGELOG](https://github.com/cbracco/cardinal/blob/master/CHANGELOG.md)

# Table of Contents

- [What is Cardinal?](#what-is-cardinal)
- [Features](#features)
- [Dependencies](#dependencies)
- [Getting Started](#getting-started)
  - [Package Managers](#package-managers)
    - [Updating Cardinal](#updating-cardinal)
  - [LESS Plugin](#less-plugin)
  - [Direct Download](#direct-download)
- [Customizing](#customizing)
  - [Including Partials](#including-partials)
  - [Modifying Variables](#modifying-variables)
- [Further Documentation](#further-documentation)
- [Browser Support](#browser-support)
- [Contributing](#contributing)
- [Versioning](#versioning)
- [Credits](#credits)
- [License](#license)

# What is Cardinal?

Cardinal is a modular, “mobile-first” CSS framework built with performance and scalability in mind.

The purpose of this framework is to make it easier for front-end web developers to prototype, build, scale, and maintain CSS for responsive websites, user interfaces, and applications. Cardinal omits many aesthetic design decisions that often bog down other CSS frameworks (which are often [UI toolkits, not frameworks](https://speakerdeck.com/csswizardry/what-is-a-css-framework-anyway)), leaving the design and creativity up to you.

# Features

- “Mobile-first” CSS written for the [LESS preprocessor](http://lesscss.org/)
- [normalize.css](https://github.com/necolas/normalize.css)
- Sensible global styles, variables, and useful mixins
- A flexible grid system and encapsulated styles for common UI objects (buttons, forms, tables, etc.)
- A collection of utility classes optimized for great gzip compression
- The entire framework, minified and gzipped, clocks in at a mere **11.27kB**

# Dependencies

If you would like to install Cardinal with the [Bower](http://bower.io/) package manager tool, you will first need to install [Node.js](http://nodejs.org/) on your machine or server. Once you have Node.js installed, make sure npm is installed ([info here](http://blog.npmjs.org/post/85484771375/how-to-install-npm)), and use it to install Bower, like so:

```
npm install -g bower
```

Also, Cardinal depends on the awesome tool [Autoprefixer](https://github.com/postcss/autoprefixer), which adds the proper vendor prefixes that provide the [specified browser support](#browser-support) for this project. This way, you don’t have to worry about manually including the correct prefixes while writing LESS/CSS. You are encouraged to incorporate Autoprefixer into your build process (this is easy to do with tools like [Grunt](http://gruntjs.com/) or [Gulp](http://gulpjs.com/)).

# Getting Started

There are several ways that you can use Cardinal in your new and existing web projects.

## Package Managers

The fastest way to get up-and-running with Cardinal is by using a package manager, like so:

**Bower**

To install Cardinal with [Bower](http://bower.io/), run the following command(s) from your terminal:

```
cd your-project-directory
bower install --save cardinal
```
Once installed, you can find Cardinal here:

```
your-project-directory
│
└───bower_components
    │
    └───cardinal
```

**npm**

To install Cardinal with [npm](https://www.npmjs.com), run the following command(s) from your terminal:

```
cd your-project-directory
npm install --save cardinalcss
```
Once installed, you can find Cardinal here:

```
your-project-directory
│
└───node_modules
    │
    └───cardinalcss
```

### Updating Cardinal

You can grab the latest version of Cardinal by running the following command(s):

**Bower**

```
bower update cardinal
```

**npm**

```
npm update cardinal
```

## LESS Plugin

There is also a [LESS.js plugin available](https://github.com/bassjobsen/less-plugin-cardinal) for Cardinal, created and maintained by [@bassjobsen](https://github.com/bassjobsen). You can use this plugin to include Cardinal in your project before your custom project code.

**Installation**

```
npm install less
npm install less-plugin-cardinal
```

**Usage**

```
lessc custom.less --cardinal
```

## Direct Download

If you do not want to use a package manager to integrate Cardinal into your project, you can also download it directly:

**Build from Source**

To do so, you need to already have Node.js and npm installed on your machine. Cardinal uses Gulp to build and watch the CSS files.

```
git clone git@github.com:cbracco/cardinal.git
cd cardinal
npm install
gulp
```

**Download the Zip**

You can also download the `.zip` file and proceed however you wish. Include the minified CSS file in the `<head>` of your project, or modify the LESS files directly (not recommended, since it will be difficult to update to future versions). It’s your world!

- [Download the zip](https://github.com/cbracco/cardinal/archive/master.zip)

# Customizing

## Including Partials

The easiest way to include Cardinal in your project is to include the entire thing before your project files, like so:

```
//
// Vendor
//

@import url("../../path/to/bower_components/cardinal/cardinal.less");

//
// Project
//

@import url("project-file.less");
```

However, you don’t have to use the entire Cardinal framework. The only required partials are located in the `base/` directory:

```
//
// Vendor
//

//
// Base
//
// These variables, mixins, and styles are required if you want to use this
// framework, and they should be included in the following order.
//

// Base -> Variables

@import url("base/variables.less");

// Base -> Mixins

@import url("base/mixins/to-rem.less");
@import url("base/mixins/media-queries.less");
@import url("base/mixins/antialiasing.less");
@import url("base/mixins/clearing.less");
@import url("base/mixins/font-face.less");
@import url("base/mixins/momentum-scrolling.less");
@import url("base/mixins/text-truncate.less");
@import url("base/mixins/vertically-centered.less");

// Base -> Normalize

@import url("base/normalize.less");

// Base -> Reset

@import url("base/reset.less");

// Base -> Debug (DO NOT USE THIS STYLESHEET IN PRODUCTION!)

//@import url("base/debug.less");

// Base -> Default Styles

@import url("base/root.less");
@import url("base/anchors.less");
@import url("base/text-elements.less");
@import url("base/headings.less");
@import url("base/lists.less");
@import url("base/horizontal-rules.less");
@import url("base/blockquotes.less");
@import url("base/code.less");
@import url("base/embedded-content.less");
@import url("base/form-elements.less");
@import url("base/tables.less");
@import url("base/print.less");
```

Beyond that, you can use whatever grid system, components, and utilities you would like. The `base/` directory just sets some good base styles for the framework.

## Modifying Variables

After getting started and including Cardinal, you will probably want to begin customizing it to suit your project. Cardinal has a bunch of LESS variables located in the `base/variables.less` file that can be tweaked to your needs. **These variables should not be edited directly.** The recommended way to make changes to Cardinal LESS variables is to create your own project-specific variables file (e.g. `project-variables.less`), and override the default variables in this file, like so:

```
// project-variables.less

// Typography -> Font Sizes -> Base

$font-size: 17;

//
// Typography -> Line height
//

$line-height: 26;
```

This file should be included in your main LESS file after you’ve included Cardinal.

# Further Documentation

There is a `README.md` file included in each major directory of this project that further details the code located in those directories.

- [Base](https://github.com/cbracco/cardinal/tree/master/base)
- [Components](https://github.com/cbracco/cardinal/tree/master/components)
- [Layout](https://github.com/cbracco/cardinal/tree/master/layout)
- [Utilities](https://github.com/cbracco/cardinal/tree/master/utilities)

# Browser support

Cardinal supports most modern browsers:

* Google Chrome 25+
* Mozilla Firefox 19+
* Safari 6.0+
* iOS Safari 6.0+
* Opera 12.1+
* Android 4.2+
* Internet Explorer 9+

# Contributing

Is something broken? Do you have ideas or feature requests? Please [submit an issue](https://github.com/cbracco/cardinal/issues/new) here on GitHub. Or, if you are feeling especially motivated, write some code and submit a pull request!

**IMPORTANT**: Before doing a bunch of work, please consult the [CONTRIBUTING.md document](https://github.com/cbracco/cardinal/blob/master/CONTRIBUTING.md), which gives detailed instructions on how to contribute to this project. If you do not follow the instructions, your code will probably not make it into this project.

# Versioning

Cardinal will be maintained using the [Semantic Versioning](http://semver.org/) guidelines. From here on out, releases will be numbered using the following format:

`major.minor.patch`

* Breaking backwards compatibility increments `major`, while resetting `minor` and `patch`.
* New code that does not break backwards compatibility increments `minor`, while resetting `patch`.
* Bug fixes and other small changes increment `patch`.

# Credits

Cardinal is a project by [@cbracco](http://twitter.com/cbracco). Its principles and techniques are very influenced by the following projects, organizations, and individuals:

- [OOCSS](http://oocss.org/in) ([@stubbornella](https://github.com/stubbornella/))
- [inuit.css](https://github.com/inuitcss/) ([@csswizardry](https://github.com/csswizardry/))
- [SUIT CSS](https://suitcss.github.io/) ([@necolas](https://github.com/necolas/))
- [TACHYONS](http://tachyons.io/) ([@mrmrs](https://github.com/mrmrs/))
- [Basscss](http://www.basscss.com/) ([@jxnblk](https://github.com/jxnblk/))
- [Scally](https://github.com/chris-pearce/scally/) ([@chris-pearce](https://github.com/chris-pearce/))
- [Atomic CSS](http://acss.io/) ([@yahoo](https://github.com/yahoo/))
- [HTML5 Boilerplate](https://github.com/h5bp/html5-boilerplate/) ([@h5bp](https://github.com/h5bp/))
- [Kraken](https://github.com/cferdinandi/kraken/) ([@cferdinandi](https://github.com/cferdinandi/))

...and any others I may have forgotten &mdash; let me know if I have, and I will certainly add them. 

Many thanks for your hard work and great ideas!

# License

Cardinal is licensed under the MIT Open Source license. For more information, see the [LICENSE.md](https://github.com/cbracco/cardinal/blob/master/LICENSE.md) file in this repository.


CHANGELOG.md
# 3.7.0 (March 24, 2016)

- Add `.breakpoint-prefixes()` mixin (thanks @lukechilds!).

# 3.6.1 (February 10, 2016)

- Fix `@font-face` mixin.

# 3.5.1 (January 18, 2016)

- Remove `outline: 0 none;` from form components.

# 3.5.0 (January 18, 2016)

- Hardcode the value of `@to-rem-font-size` to 16 so that `@font-size` can be overridden.

# 3.4.1 (January 11, 2016)

- Fix incorrect variable used in the grid system.
- Fix `height: auto !important;` from preventing `<textarea>` resizing in latest Chrome, Safari, and Firefox.

# 3.3.1 (January 7, 2016)

- Add back `max-width` mixins removed in 3.3.0 because they caused major breaking changes.
- Add deprecation messages to `max-width` mixins (thanks @lukechilds).
- Deprecate v3.3.0.

# 3.3.0 (December 21, 2015)

- Remove mixins that are not aligned with the “mobile-first” methodology (thanks @mrmrs for calling me out).

# 3.2.9 (December 19, 2015)

- fix incorrect URL (typo) in `README.md`.

# 3.2.8 (December 19, 2015)

- Had to bump version number because I pushed 3.2.7 to npm too soon, and had to unpublish it since version numbers in the CSS files were incorrect.

# 3.2.7 (December 19, 2015)

- Adding credit where credit is due. I cannot believe I forgot to do this (thank you @mrmrs for calling me out).
- Updated code comments.
- Removed some commented-out code.
- Add npm/Bower badges to `README.md`.

# 3.2.6 (September 29, 2015)

- Update `gulp-less` to latest version.
- Update `gulpfile.js`.

# 3.2.5 (August 5, 2015)

- Update some variables to use the `@spacing-*` equivalent value set previously, instead of hard-coded integers.

# 3.2.4 (May 28, 2015)

- Remove `text-rendering` property from headings ([more info](http://cbrac.co/1Fjv7IR)).
- Update `.gitignore` (exclude sourcemaps from repo).

# 3.2.3 (May 15, 2015)

- Update `README.md` to include customization docs.

# 3.2.2 (May 12, 2015)

- Add [Codepen demos](http://codepen.io/collection/nmrWQR/) to documentation.
- Fix typos in `index.html`.

# 3.2.1 (May 12, 2015)

- Add more documentation in the form of `README.md` files.
- Clean up `:focus` styles, move declarations to `root.less`.

# 3.2.0 (May 11, 2015)

- Remove `::selection` styles because they cause iOS 8 to crash in certain scenarios.
- Simplify directory structure for LESS partials
- Add `.list-reset` component modifier.

# 3.1.0 (March 8, 2015)

- Remove `@root-font-family`, `@root-font-size`, and `@root-line-height` variables because they were causing problems. Using `@font-family-*`, `@font-size`, and `@line-height` variables instead as root.
- Use `@font-size-0` variable in `layout/grids.less` instead of `font-size: 0;`

# 3.0.7 (March 8, 2015)

- Fix `.btn-round` bug for `input[type="button"]` and `input[type="submit"]` elements.

# 3.0.6 (February 25, 2015)

- Remove default margin for `.grid-item` elements on `li` elements (thanks, @bassjobsen!).

# 3.0.5 (February 18, 2015)

- Add `!important` to utility classes.

# 3.0.4 (February 16, 2015)

- Fix `@print` typo in `layout/wrappers.less`.

# 3.0.3 (February 9, 2015)

- Add `:disabled` to disabled button styles.

# 3.0.2 (February 9, 2015)

- Update `README.md`.
- Comment out `core/debug.less` by default.

# 3.0.1 (February 5, 2015)

- Update `bower.json` and `README.md`.

# 3.0.0 (February 5, 2015)

- Overhauled Cardinal (huzzah!) to function as a CSS framework only.
  - Removed all "boilerplate" files and folders.
  - Replaced Grunt with Gulp.
  - Moved `test.html` to `index.html`.
  - Updated dotfiles, `bower.json`, and `package.json`.
  - Moved CSS and LESS files up a level, beginning in the root directory.
  - Added a `.csscomb.json` file.
  - Changed maximum line length from 72 to 80 characters.
  - Added a bunch of useful utility classes.
  - Added a simple `debug.less` file to use while developing.
- Updated `README.md`.

# 2.4.5 (November 7, 2014)
- Removed `@viewport` CSS rules because they break Google Chrome’s "device mode" tool. They will be added back when they are needed, or when the bug has been fixed (whichever comes first).

# 2.4.4 (August 14, 2014)

- Added whole gutter styles that were missing in previous versions (via @nastasiu).

# 2.4.3 (August 13, 2014)

- Fixed parent grid gutter styles being applied to nested child grids that may need different gutter styles (via @nastasiu).

# 2.4.2 (July 31, 2014)

- Added missing width classes in `layout.less` (via @nastasiu).

# 2.4.1 (July 31, 2014)

- Added `node_modules/` to the `.gitignore` file (via @brandonb927).
- Added missing grunt dependency to `package.json` file (via @brandonb927).

# 2.4.0 (May 15, 2014)

- Update jQuery to 2.1.1.
- Update Modernizr to 2.8.1.

# 2.3.2 (April 9, 2014)

- Add stylesheet to `test.html`.
- Remove `<nav>` element from `test.html`.

# 2.3.1 (April 8, 2014)

- Updated `test.html` because it was missing many tests. Not sure how they disappeared, but they are back now.

# 2.3.0 (April 7, 2014)

- Added my [html5-test-page](https://github.com/cbracco/html5-test-page) so you can see how some common HTML elements react to Cardinal (thanks [@petsagouris](https://github.com/petsagouris) for getting this going).
- Added a `CONTRIBUTING.md` file with stricter guidelines on how to contribute in hopes that people will actually read it before submitting pull requests.
- Updated `README.md` to mention the new `CONTRIBUTING.md` file.
- Moved uncompressed assets into their own directory.
- Updated `.gitignore` file to ignore new `/dist` directory.
- Cleaned up `gruntfile.js`. It was missing some semicolons, and the default and watch tasks were separated (thanks [@petsagouris](https://github.com/petsagouris)).
- Added `@button-border-radius` LESS variable (thanks [brandonb927](https://github.com/brandonb927)).
- Added `font-weight: 700;` to buttons by default (thanks [brandonb927](https://github.com/brandonb927)).

# 2.2.6 (March 27, 2014)

- Updated normalize.css to 3.0.1.

# 2.2.5 (February 18, 2014)


js/main.js
// When the DOM is ready,
$(function() {

    /**
     * Google Analytics event click tracking
     */

    $("#header-download").on("click", function() {
        ga("send", "event", "Header actions", "click", "Download");
    });

    $("#header-view").on("click", function() {
        ga("send", "event", "Header actions", "click", "GitHub Repo");
    });

    $("#header-changelog").on("click", function() {
        ga("send", "event", "Header actions", "click", "Changelog");
    });

    $("#header-readme").on("click", function() {
        ga("send", "event", "Header actions", "click", "Readme");
    });

    $("#footer-cbracco").on("click", function() {
        ga("send", "event", "Footer actions", "click", "cbracco.me");
    });

    /**
     * Request the latest version number from GitHub once a minute
     */

    setInterval(updateVersion(), 60 * 1000);

});


gulpfile.js
// Load dependencies
var browserSync = require('browser-sync')
    browserSyncReload = browserSync.reload,
    del = require('del'),
    gulp = require('gulp'),
    gulpLoadPlugins = require('gulp-load-plugins'),
    plugins = gulpLoadPlugins();

// Set browser support for gulp-autoprefixer
var browserSupport = [
  'last 2 versions',
  'Android >= 4.2',
  'Chrome >= 25',
  'Firefox >= 19',
  'ie >= 9',
  'iOS >= 6',
  'Opera >= 12.1',
  'Safari >= 6'
];

// Set paths
var paths = {
  css: {
    dir: './',
    main: './cardinal.css',
    all: './*.css'
  },
  less: {
    dir: './',
    main: './cardinal.less',
    all: './**/*.less'
  }
};

// Set banner
var pkg = require('./package.json'),
    banner = ['/*! <%= pkg.name %> v<%= pkg.version %> | <%= pkg.license %> License | <%= pkg.homepage %> */\n'];

// Error handler
var onError = function(err) {
  plugins.util.beep(); // Trigger a system beep
  plugins.util.log(
    plugins.util.colors.underline(plugins.util.colors.red('ERROR:'))
    + ' '
    + plugins.util.colors.cyan(err.plugin)
    + ' - '
    + err.message
  ); // Log a readable error message
};

// Clean (gulp clean)
gulp.task('clean', function(cb) {
  del([paths.css.dir], cb);
});

// Styles (gulp styles)
gulp.task('styles', function() {
  return gulp.src([
    paths.less.main
  ])
  .pipe(plugins.plumber({
    errorHandler: onError
  }))
  .pipe(plugins.sourcemaps.init({
    loadMaps: true
  }))
  .pipe(plugins.less())
  .pipe(plugins.autoprefixer({
    browsers: browserSupport
  }))
  .pipe(plugins.header(banner, { pkg: pkg }))
  .pipe(plugins.sourcemaps.write("./"))
  .pipe(gulp.dest(paths.css.dir))
  .pipe(plugins.csscomb())
  .pipe(plugins.size({
    gzip: false,
    showFiles: true,
    title:'unminified css'
  }))
  .pipe(plugins.minifyCss({
    shorthandCompacting: false
  }))
  .pipe(plugins.size({
    gzip: false,
    showFiles: true,
    title:'minified css'
  }))
  .pipe(plugins.size({
    gzip: true,
    showFiles: true,
    title:'minified css'
  }))
  .pipe(plugins.rename({
    suffix: '.min'
  }))
  .pipe(gulp.dest(paths.css.dir))
  .pipe(browserSyncReload({
    stream: true
  }));
});

// Browser Sync (gulp browser-sync)
gulp.task('browser-sync', function() {
  browserSync({
    server: {
      baseDir: "./"
    }
  });
});

// Reload all Browsers
gulp.task('bs-reload', function () {
    browserSyncReload();
});

// Default (gulp)
gulp.task('default', ['styles', 'browser-sync'], function() {
  gulp.watch(paths.less.all, ['styles']);
  gulp.watch('*.html', ['bs-reload']);
});


Gruntfile.js
module.exports = function(grunt) {
    'use strict'

    var path = require('path')

    // Force use of Unix newlines
    grunt.util.linefeed = '\n'

    grunt.initConfig({
        pkg : grunt.file.readJSON('package.json'),

        banner: '/*!\n' +
                ' * Name: Cardinal v<%= pkg.version %> (<%= pkg.homepage %>)\n' +
                ' * Info: <%= pkg.description %>\n' +
                ' * Copyright <%= grunt.template.today("yyyy") %> <%= pkg.author %>\n' +
                ' * Licensed under <%= _.pluck(pkg.licenses, "type") %> (<%= _.pluck(pkg.licenses, "url") %>)\n' +
                ' */\n',

        // Directory variables
        lessDir : 'less/',
        cssDir  : 'css/',
        jsDir  : 'js/',

        // Run LESS CSS compilation
        less: {
            compile: {
                files: {
                    "<%= cssDir %><%= pkg.name %>.css": "<%= lessDir %><%= pkg.name %>.less"
                }
            },
            minify: {
                options: {
                    cleancss: true,
                    report: 'min'
                },
                files: {
                    'css/<%= pkg.name %>.min.css': 'css/<%= pkg.name %>.css',
                }
            }
        },

        // Automatically prefix CSS based on Can I Use database
        autoprefixer: {
            options: {
                browsers: [
                    'chrome 25',
                    'firefox 19',
                    'safari 6',
                    'ie 9',
                    'opera 12.1',
                    'ios 4',
                    'android 4.2'
                ]
            },
            files: {
                src: [
                    'css/<%= pkg.name %>.css',
                    'css/<%= pkg.name %>.min.css',
                ]
            }
        },

        // Remove comments final CSS
        cssmin: {
            options: {
                keepSpecialComments: 0
            },
            single_file: {
                src: 'css/<%= pkg.name %>.min.css',
                dest: 'css/<%= pkg.name %>.min.css'
            }
        },

        // Add a nice banner to the compiled CSS
        usebanner: {
            options: {
                position: 'top',
                banner: '<%= banner %>'
            },
            files: {
                src: [
                    'css/<%= pkg.name %>.css',
                    'css/<%= pkg.name %>.min.css',
                ]
            }
        },

        // Watch for changes on these files and recompile when changed
        watch: {
            options: {
                nospawn: true
            },
            less: {
                files: ['<%= lessDir %>**/*.less'],
                tasks: ['less', 'usebanner']
            }
        }
    })


    // Grunt plugins
    grunt.loadNpmTasks('grunt-autoprefixer')
    grunt.loadNpmTasks('grunt-banner')
    grunt.loadNpmTasks('grunt-contrib-cssmin')
    grunt.loadNpmTasks('grunt-contrib-less')
    grunt.loadNpmTasks('grunt-contrib-watch')

    // Registered Grunt tasks
    grunt.registerTask('default', [
        'less',
        'autoprefixer',
        'cssmin',
        'usebanner',
        'watch'
    ])
}


js/plugins.js
/**
 * Avoid `console` errors in browsers that lack a console.
 */

(function() {
    var method;
    var noop = function noop() {};
    var methods = [
        'assert', 'clear', 'count', 'debug', 'dir', 'dirxml', 'error',
        'exception', 'group', 'groupCollapsed', 'groupEnd', 'info', 'log',
        'markTimeline', 'profile', 'profileEnd', 'table', 'time', 'timeEnd',
        'timeStamp', 'trace', 'warn'
    ];
    var length = methods.length;
    var console = (window.console = window.console || {});

    while (length--) {
        method = methods[length];

        // Only stub undefined methods.
        if (!console[method]) {
            console[method] = noop;
        }
    }
}());


/**
 * Grab the latest version tag from GitHub and append it
 */

function updateVersion() {

    $.ajax({
        url: "https://api.github.com/repos/cardinalcss/cardinalcss/tags",
        dataType: "jsonp",
        timeout: 60 * 1000,
        success: function(results) {
            var tag = results.data;
            $(".js-version").each(function() {
                $(this).append("&nbsp;<span class=\"version\">" + tag[0]["name"] + "</span>");
            });
        }
    });

}


gruntfile.js
module.exports = function(grunt) {
    'use strict';

    var path = require('path');

    // Force use of Unix newlines
    grunt.util.linefeed = '\n';

    grunt.initConfig({
        pkg: grunt.file.readJSON('package.json'),
        pkgName: '<%= pkg.name.toLowerCase() %>',
        banner: '/*! <%= pkgName %> v<%= pkg.version %> | <%= _.pluck(pkg.licenses, "type") %> License | <%= pkg.homepage %> */',

        // Directory variables
        cssDir: 'css/',
        distDir: 'dist/',
        jsDir: 'js/',
        lessDir: 'less/',
        sassDir: 'sass/',

        // Run LESS CSS compilation
        less: {
            compile: {
                files: {
                    '<%= distDir %><%= cssDir %><%= pkgName %>.css': '<%= lessDir %><%= pkgName %>.less'
                }
            },
            minify: {
                options: {
                    cleancss: true,
                    report: 'min'
                },
                files: {
                    '<%= distDir %><%= cssDir %><%= pkgName %>.min.css': '<%= distDir %><%= cssDir %><%= pkgName %>.css'
                }
            }
        },

        // Run SASS CSS compilation
        sass: {
            compile: {
                files: {
                    '<%= distDir %><%= cssDir %><%= pkgName %>.sass.css': '<%= sassDir %><%= pkgName %>.scss'
                }
            },
            minify: {
                options: {
                    style: 'compressed'
                },
                files: {
                    '<%= distDir %><%= cssDir %><%= pkgName %>.sass.min.css': '<%= distDir %><%= cssDir %><%= pkgName %>.sass.css'
                }
            }
        },

        // Automatically prefix CSS based on Can I Use database for the following browsers versions
        autoprefixer: {
            development: {
                browsers: [
                    'chrome 25',
                    'firefox 19',
                    'safari 6',
                    'ie 9',
                    'opera 12.1',
                    'ios 4',
                    'android 4.2'
                ],
                files: {
                    '<%= distDir %><%= cssDir %><%= pkgName %>.css': '<%= distDir %><%= cssDir %><%= pkgName %>.css'
                }
            }
        },

        copy: {
            dist: {
                files: [
                    {
                        expand: true,
                        src: ['<%= jsDir %>**/*.js'],
                        dest: '<%= distDir %>'
                    }
                ]
            }
        },

        // Add a nice banner to the CSS files
        usebanner: {
            options: {
                position: 'top',
                banner: '<%= banner %>'
            },
            src: '<%= distDir %><%= cssDir %>**/*.css',
            dest: '<%= distDir %>'
        },

        // Watch for changes on these files and recompile when changed
        watch: {
            less: {
                files: ['<%= lessDir %>**/*.less'],
                tasks: ['default']
            },
            sass: {
                files: ['<%= sassDir %>**/*.scss'],
                tasks: ['default']
            }
        }
    });

    // Grunt plugins
    grunt.loadNpmTasks('grunt-autoprefixer');
    grunt.loadNpmTasks('grunt-banner');
    grunt.loadNpmTasks('grunt-contrib-copy');
    grunt.loadNpmTasks('grunt-contrib-less');
    grunt.loadNpmTasks('grunt-contrib-sass');
    grunt.loadNpmTasks('grunt-contrib-watch');

    // Registered Grunt tasks
    grunt.registerTask('default', [
        'less:compile',
        'sass:compile',
        'autoprefixer',
        'less:minify',
        'sass:minify',
        'usebanner',
        'watch'
    ]);

    grunt.registerTask('dist', [
        'less:compile',
        'sass:compile',
        'autoprefixer',
        'less:minify',
        'sass:minify',
        'usebanner',
        'copy:dist'
    ]);
};
